### PR TITLE
Fix DataLoader inspection and re-instantiation in Lite

### LINKED ISSUE #10329: DataLoader wrapping, re-instantiation and patching
## 🐛 Bug

Recent changes to the Lite DataLoader in #10279 have not been properly reviewed and tested. Several clean ups need to be done and better tests need to be written. 

The following issue exists:

### To Reproduce

```python
def test_setup_custom_dataloaders_new():

    class DataLoaderSubclass1(DataLoader):
        def __init__(self, attribute1, *args, **kwargs):
            # UNCOMMENT THE LINE BELOW, NO DIFFERENCE (EXPECTED)
            # self.attribute1 = attribute1
            super().__init__(*args, **kwargs)

    class DataLoaderSubclass2(DataLoaderSubclass1):
        def __init__(self, attribute1, attribute2, *args, **kwargs):
            # UNCOMMENT THE LINE BELOW, TEST PASSES
            # self.attribute2 = attribute2
            super().__init__(attribute1, *args, **kwargs)

    class LiteWithCustomDataLoader(LightningLite):
        def run(self):
            dataloader = DataLoaderSubclass2("attribute1", "attribute2", dataset=range(4), batch_size=2)
            lite_dataloader = self.setup_dataloaders(dataloader)
            
            # THIS ASSERTION PASSES
            assert lite_dataloader._dataloader.attribute1 == "attribute1"
            
            # THIS ASSERTION FAILS
            assert lite_dataloader._dataloader.attribute2 == "attribute2"

    LiteWithCustomDataLoader().run()
```

### Expected behavior

According to the changes in the PR #10279, this is supposed to work and was the main motivation of it. But it does not because the tests were insufficient. This issue was found by me in #10297 while trying to clean up the code. 

### Environment

Lightning v1.5.0!

cc @tchaton @rohitgr7 @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #10329

A detailed explanation of why the bug occurs can be found in the linked issue.
A newly added test demonstrates the failure on master. 

This fix should be reviewed carefully, at least by @carmocca and @tchaton .

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

**Review by tchaton — APPROVED** [review]

**Comment by tchaton** [comment]
@carmocca Do you think we should bring this to Lightning too ?

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -86,6 +86,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `apply_to_collection(defaultdict)` ([#10316](https://github.com/PyTorchLightning/pytorch-lightning/issues/10316))
 
 
+- Fixed interception of ``__init__`` arguments for sub-classed DataLoader re-instantiation in Lite ([#10334](https://github.com/PyTorchLightning/pytorch-lightning/issues/10334))
```
Comment (carmocca) [inline CHANGELOG.md:89]:
```suggestion
- Fixed interception of `__init__` arguments for sub-classed DataLoader re-instantiation in Lite ([#10334](https://github.com/PyTorchLightning/pytorch-lightning/issues/10334))
```

**Review by carmocca — APPROVED** [review]
LGTM

**Comment by awaelchli** [comment]
Is it the approach of how we handle dataloaders in general or the fix that I added?
